active/0021-transfer-files-over-mqtt.md

```diff
@@ -0,0 +1,175 @@
+# MQTT based file transfer
+
+## Changelog
+
+* 2022-09-20: @qzhuyan, @id Initial draft
+
+## Abstract
+
+TODO:
+
+## Motivation
+
+EMQX customers are asking for file transfer functionality from IoT devices to their cloud (primary use case), and from cloud to IoT devices over MQTT. Right now they are uploading files from devices via FTP or HTTPS (e.g. to S3), but it's not working well. For example, both FTP and HTTP servers usually struggle to keep with large number of simultaneous bandwith-intensive connections. Another issue is that uploading files over weak/spotty/unreliable network is not a good experience due to frequent connection reset and transfer restarts.
+
+Device-to-cloud file transfer use cases:
+
+* [CAN bus](https://en.wikipedia.org/wiki/CAN_bus) data
+* Image taken by industry camera for Quality Assurance
+* Large data file collected from forklift
+* Video and audio data from truck cars, and video data captured by inbound unloading cameras
+* Vehicle real-time logging data and messaging
+* ML logs
+
+Cloud-to-device file transfer use cases:
+
+* AI/ML models
+* Firmware upgrades
+
+With EMQX, Customers can transfer real-time IoT data (e.g. structured, as well as unstructured IoT, sensor, and industrial data), and will also support offline types of bulk data transfer (video, audio, images,  compressed log files, etc.), which, combined with EMQX's rules engine, will make it extremely easy for users to connect any IoT data to the cloud.
+
+## Requirements
+
+* Simultaneous transfer of MQTT and various types of files over a single network connection
+* Separate data transfer and file transfer channels to achieve non-blocking transmission
+  * Channel priority setting, specify the priority of different channels according to customer requirements
+  * Channels can cooperate with each other to achieve overall flow control
+  * Multiple channels can be specified for file transfer at the same time to increase file transfer parallelism
+* Efficient transmission support in weak network
+  * File transfer in weak networks such as Telematics and mobile
+  * Automatic file chunking, block-based granularity breakpoint transfer (or file-level breakpoint transfer?)
```

Comment:
Can the block size configured by user? For example, 1024 is the default block size, and user can set to 2048 etc.

Reply:
I don't see why not.